pkg/runtime/kubeadm_runtime.go

```diff
@@ -0,0 +1,200 @@
+	if err := mergo.Merge(hostSSH, k.Cluster.Spec.SSH); err != nil {
+		return nil, err
+	}
+
+	return ssh.NewSSHClient(hostSSH), nil
+}
+
+func (k *KubeadmRuntime) checkList() error {
+	return k.checkIPList()
+}
+
+func (k *KubeadmRuntime) checkIPList() error {
+	if len(k.Spec.Hosts) < 1 {
+		return fmt.Errorf("master hosts should not < 1, hosts len is %s", len(k.Spec.Hosts))
+	}
+	if len(k.Spec.Hosts[0].IPS) < 1 {
+		return fmt.Errorf("master hosts ip should not < 1, hosts ip len is %s", len(k.Spec.Hosts[0].IPS))
+	}
+	return nil
+}
+
+func (k *KubeadmRuntime) getClusterName() string {
+	return k.Cluster.Name
+}
+
+func (k *KubeadmRuntime) getHostSSHClient(hostIP string) (ssh.Interface, error) {
+	for _, host := range k.Cluster.Spec.Hosts {
+		for _, ip := range host.IPS {
+			if hostIP == ip {
+				return k.getSSHClient(&host.SSH)
+			}
+		}
+	}
+	return nil, fmt.Errorf("get host ssh client failed, host ip %s not in hosts ip list", hostIP)
+}
+
+func (k *KubeadmRuntime) getRootfs() string {
+	return common.DefaultTheClusterRootfsDir(k.Cluster.Name)
+}
+
+func (k *KubeadmRuntime) getBasePath() string {
+	return path.Join(common.DefaultClusterRootfsDir, k.Cluster.Name)
+}
+
+func (k *KubeadmRuntime) getMaster0IP() string {
+	// already check ip list when new the runtime
+	return k.Cluster.Spec.Hosts[0].IPS[0]
+}
+
+func (k *KubeadmRuntime) getDefaultKubeadmConfig() string {
+	return filepath.Join(k.getRootfs(), "etc", "kubeadm.yml")
+}
+
+func (k *KubeadmRuntime) getCloudImageDir() string {
+	return filepath.Join(common.DefaultMountCloudImageDir(k.Cluster.Name))
+}
+
+func (k *KubeadmRuntime) getCertPath() string {
+	return path.Join(common.DefaultClusterRootfsDir, k.Cluster.Name, "pki")
+}
+
+func (k *KubeadmRuntime) getEtcdCertPath() string {
+	return path.Join(common.DefaultClusterRootfsDir, k.Cluster.Name, "pki", "etcd")
+}
+
+func (k *KubeadmRuntime) getStaticFileDir() string {
+	return path.Join(k.getRootfs(), "statics")
+}
+
+func (k *KubeadmRuntime) getSvcCIDR() string {
+	return k.ClusterConfiguration.Networking.ServiceSubnet
+}
+
+func (k *KubeadmRuntime) getDNSDomain() string {
+	return k.ClusterConfiguration.Networking.DNSDomain
+}
+
+func (k *KubeadmRuntime) getAPIServerDomain() string {
+	return k.RuntimeConfig.APIServerDomain
+}
+
+func (k *KubeadmRuntime) getKubeVersion() string {
+	return k.KubernetesVersion
+}
+
+func (k *KubeadmRuntime) getVIP() string {
+	return DefaultVIP
+}
+
+func (k *KubeadmRuntime) getMasterIPList() (masters []string) {
+	return k.getHostsIPByRole(common.MASTER)
+}
+
+func (k *KubeadmRuntime) getNodesIPList() (nodes []string) {
+	return k.getHostsIPByRole(common.NODE)
+}
+
+func (k *KubeadmRuntime) getHostsIPByRole(role string) (nodes []string) {
+	for _, host := range k.Spec.Hosts {
+		if utils.InList(role, host.Roles) {
+			nodes = append(nodes, host.IPS...)
+		}
+	}
+
+	return
+}
+
+func (k *KubeadmRuntime) WaitSSHReady(tryTimes int, hosts ...string) error {
+	var err error
```

Comment:
I am wondering if we need to add a lock to protect the race section since several goroutine would read/write the global `err`.

Reply:
The better way is to use an error channel.